### PR TITLE
8275718: Relax memory constraint on exception counter updates

### PR DESCRIPTION
**Pull request by zhengyu123**
This is another instance of counter updates that only need atomic guarantee.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8275718](https://bugs.openjdk.java.net/browse/JDK-8275718): Relax memory constraint on exception counter updates

### Reviewers
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Yumin Qi](https://openjdk.java.net/census#minqi) (@yminqi - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6065/head:pull/6065` \
`$ git checkout pull/6065`

Update a local copy of the PR: \
`$ git checkout pull/6065` \
`$ git pull https://git.openjdk.java.net/jdk pull/6065/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6065`

View PR using the GUI difftool: \
`$ git pr show -t 6065`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6065.diff">https://git.openjdk.java.net/jdk/pull/6065.diff</a>

</details>

### CODE REVIEW

**Comment by dholmes-ora** [comment]
What does this achieve? These are not "hot" counters so performance is of zero concern. The conservative memory ordering doesn't guarantee cross-thread visibility but it does seem to me there is less likelihood of error reporting seeing a stale value from another thread if we have "conservative" memory ordering (which traditionally should have been a full bi-directional fence for atomic r-m-w operations).

David

**Comment by zhengyu123** [comment]
> What does this achieve? These are not "hot" counters so performance is of zero concern. The conservative memory ordering doesn't guarantee cross-thread visibility but it does seem to me there is less likelihood of error reporting seeing a stale value from another thread if we have "conservative" memory ordering (which traditionally should have been a full bi-directional fence for atomic r-m-w operations).
> 
> David

Hi David,

I agree that they are not "hot" counters, but I disagree that "conservative" memory ordering provides any benefits that prevent other threads from reading stale values.

We had internal discussion on this topic, Aleksey pointed out:
"All modifications to any particular atomic variable occur in a total order that is specific to this one atomic variable". This guarantee holds even for relaxed atomic load/stores. This is a very basic guarantee. 

In short, value updated via atomic r-m-w operation should be visible to other threads guaranteed by coherence protocol

**Comment by theRealAph** [comment]
> We had internal discussion on this topic, Aleksey pointed out: "All modifications to any particular atomic variable occur in a total order that is specific to this one atomic variable". This guarantee holds even for relaxed atomic load/stores. This is a very basic guarantee.

I think that's true for most processors as a consequence of multi-copy atomicity, but we support Power which is not multi-copy atomic, where stores can become visible to one group of threads before they become visible to all threads.

**Comment by shipilev** [comment]
(I am not arguing in favor or against this particular change, but I think we can talk a bit about generic stuff here...)

> I don't know where this guarantee is coming from. Two r-m-w atomic ops must have some guarantee via coherence for the atomic op to actually work. And an implementation could make any atomic r-m-w implementation ensure global immediate visibility. But you cannot assume this is guaranteed for all hardware. Even for a given platform this would need to be a specified guarantee in the architecture manual, not just something deduced/inferred by reasoning.

Hotspot's `memory_order_relaxed` is [aligned](https://github.com/openjdk/jdk/blob/5bb1992b8408a0d196b1afa308bc00d007458dbd/src/hotspot/share/runtime/atomic.hpp#L44-L45) with C++11 atomics semantics. C++11 atomic semantics for relaxed atomic ops requires [single modification order consistency](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering), which implies [coherence](https://en.cppreference.com/w/cpp/atomic/memory_order#Modification_order).

All known hardware platforms provide coherence out of the box (they are, indeed, cache-coherent platforms), that's why it is easy to implement in C++ (`mo_relaxed`) and in Java (`VarHandles.(get|set)opaque`).

I am always confused by "immediate global visibility". The problem with statements that include "immediate", "before", "after" is that they leak in the notion of time, which is ill-defined for a single memory location without any reference to other variables. Maybe you can expand your concern with the example?

**Comment by theRealAph** [comment]
> > We had internal discussion on this topic, Aleksey pointed out: "All modifications to any particular atomic variable occur in a total order that is specific to this one atomic variable". This guarantee holds even for relaxed atomic load/stores. This is a very basic guarantee.
> 
> I think that's true for most processors as a consequence of multi-copy atomicity, but we support Power which is not multi-copy atomic, where stores can become visible to one group of threads before they become visible to all threads.

Sorry, this was something of a red herring.

My main point: imposing ordering with respect to other memory accesses around a counter increment does nothing useful unless you care about the ordering of the increment with respect to those other accesses, which AFAICS you don't in this case.

**Comment by zhengyu123** [comment]
> _Mailing list message from [David Holmes](mailto:david.holmes@oracle.com) on [hotspot-dev](mailto:hotspot-dev@mail.openjdk.java.net):_
> 
> Correction ...
> 
> On 3/11/2021 10:09 pm, David Holmes wrote:
> 
> > On 1/11/2021 7:16 pm, Aleksey Shipilev wrote:
> > > On Thu, 21 Oct 2021 15:16:28 GMT, Zhengyu Gu <zgu at openjdk.org> wrote:
> > > > This is another instance of counter updates that only need atomic
> > > > guarantee.
> > > 
> > > 
> > > (I am not arguing in favor or against this particular change, but I
> > > think we can talk a bit about generic stuff here...)
> > > > I don't know where this guarantee is coming from. Two r-m-w atomic
> > > > ops must have some guarantee via coherence for the atomic op to
> > > > actually work. And an implementation could make any atomic r-m-w
> > > > implementation ensure global immediate visibility. But you cannot
> > > > assume this is guaranteed for all hardware. Even for a given platform
> > > > this would need to be a specified guarantee in the architecture
> > > > manual, not just something deduced/inferred by reasoning.
> > > 
> > > 
> > > Hotspot's `memory_order_relaxed` is
> > > [aligned](https://github.com/openjdk/jdk/blob/5bb1992b8408a0d196b1afa308bc00d007458dbd/src/hotspot/share/runtime/atomic.hpp#L44-L45)
> > > with C++11 atomics semantics. C++11 atomic semantics for relaxed
> > > atomic ops requires [single modification order
> > > consistency](https://en.cppreference.com/w/cpp/atomic/memory_order#Relaxed_ordering),
> > > which implies
> > > [coherence](https://en.cppreference.com/w/cpp/atomic/memory_order#Modification_order).
> > > All known hardware platforms provide coherence out of the box (they
> > > are, indeed, cache-coherent platforms), that's why it is easy to
> > > implement in C++ (`mo_relaxed`) and in Java
> > > (`VarHandles.(get|set)opaque`).
> > > I am always confused by "immediate global visibility". The problem
> > > with statements that include "immediate", "before", "after" is that
> > > they leak in the notion of time, which is ill-defined for a single
> > > memory location without any reference to other variables. Maybe you
> > > can expand your concern with the example?
> > 
> > 
> > Let me back up to be clear. I stated that memory-order-conservative
> > might lower the chances (in a general platform-agnostic way) of seeing a
> > stale value, compared to memory-order-relaxed, due to the stronger
> > memory fence/barrier operation it implies. The response to that was:
> > "value updated via atomic r-m-w operation should be visible to other
> > threads guaranteed by coherence protocol"
> > claiming that visibility guarantees were inherently present due to
> > coherence regardless of what kind of memory fence/barrier were
> > associated with the r-m-w atomic operation. I'm not sure if that is
> > actually true. If it is true then we would not need any memory-order
> > parameter on the r-m-w atomic operations because they would be all be
> > the same due to this underlying coherence property.
> 
> No that isn't true. I see now that the C++ "Modification Order" definition requires the write to the counter to be (for want of a better term) "immediately visible" to any subsequent read - so no stale value could be read. That is a far stronger guarantee than I expected from mo_relaxed. The use of other mo values on the r-m-w atomic operation impact the ordering between that variable and other atomic variables.
> 
> David -----
>
Yes, for this single location atomic counter, there is no ordering involved. Although the counters are not hot, but more restricted memory constraints do not add any values.

Are you okay with this change?

Thanks,

-Zhengyu
 
> > When I said "immediate global visibility" I was referring to a situation
> > where once the write in the r-m-w atomic op had occurred then all
> > subsequent reads would see the value of that write. It is true that such
> > a thing may not require "immediacy" in a temporal sense, but the net
> > effect is the same.
> > David
> > -----

**Review by dholmes-ora — APPROVED** [review]
I'm not sure there is any actual benefit to this change, but I also do not see any harm. So okay.

Thanks,
David

**Comment by zhengyu123** [comment]
> I'm not sure there is any actual benefit to this change, but I also do not see any harm. So okay.
> 
> Thanks, David

Thanks, @dholmes-ora 

I don't believe it has measurable impact neither. In theory, mo_conservative is much more expensive ...

-Zhengyu

**Review by yminqi — APPROVED** [review]
LGTM.

**Comment by zhengyu123** [comment]
Thanks, @yminqi
